src/Data/FunctorWithIndex.purs

```diff
@@ -0,0 +1,57 @@
+module Data.FunctorWithIndex
+  ( class FunctorWithIndex, imap
+  ) where
+
+import Prelude
+
+import Data.Maybe (Maybe)
+import Data.Maybe.First (First)
+import Data.Maybe.Last (Last)
+import Data.Monoid.Additive (Additive)
+import Data.Monoid.Conj (Conj)
+import Data.Monoid.Disj (Disj)
+import Data.Monoid.Dual (Dual)
+import Data.Monoid.Multiplicative (Multiplicative)
+
+
+-- | A `Functor` with an additional index.  
+-- | Instances must satisfy a modified form of the `Functor` laws
+-- | ```purescript
+-- | imap (\_ a -> a) = id
+-- | imap f . imap g = imap (\i -> f i <<< g i)
+-- | ```
+-- | and be compatible with the `Functor` instance
+-- | ```purescript
+-- | map f = imap (const f)
+-- | ```
+class Functor f <= FunctorWithIndex i f | f -> i where
+  imap :: forall a b. (i -> a -> b) -> f a -> f b
```

Comment:
This is called `mapWithIndex` [elsewhere](https://pursuit.purescript.org/packages/purescript-arrays/4.1.1/docs/Data.Array#v:mapWithIndex) actually.